Software/Development/SemesterProject/App/encoderManager.c

```diff
@@ -0,0 +1,211 @@
+{
+    motorPosCalc();
+    motorSpeedCalc();
+    clearInterruptFlag();
+}
+
+
+float obtainDeltaTheta(void)
+{
+    //Calculate delta and check if it has jumped to next lap.
+    float deltaTheta;
+    deltaTheta = fabs(motorPosSpeedObject.thetaElec - motorPosSpeedObject.thetaElecOld);
+    if (deltaTheta > PI) deltaTheta = fabs(TWO_PI - deltaTheta);
+    return deltaTheta;
+}
+
+
+void updateSpeed(float deltaTheta)
+{
+    //Check direction and calculate electrical frequency.
+    if (motorPosSpeedObject.dir == 0) motorPosSpeedObject.freqElec = deltaTheta * RAD_TO_REV * 1000000 / motorPosSpeedObject.speedTempCount;
+    else motorPosSpeedObject.freqElec = - deltaTheta * RAD_TO_REV * 1000000 / motorPosSpeedObject.speedTempCount; //Going reverse direction.
+
+    //Update other values.
+    motorPosSpeedObject.thetaElecOld = motorPosSpeedObject.thetaElec;
+    motorPosSpeedObject.speedTempCount = 0;
+}
+
+
+void calcOtherSpeeds(void)
+{
+    motorPosSpeedObject.freqMech = motorPosSpeedObject.freqElec * POLE_PAIRS_INVERSE;    //Hz
+    motorPosSpeedObject.rpmMech = (int16) (motorPosSpeedObject.freqMech * 60.0);    //To go from Hz to rpm.
+}
+
+
+/*
+ * Function that updates the object containing position info from encoder.
+ */
+void motorPosCalc(void)
+{
+    // Check motor direction: 0=CCW/reverse, 1=CW/forward.
+    motorPosSpeedObject.dir = EQep1Regs.QEPSTS.bit.QDF;
+
+    //Update raw angle with counter, remember that QPOSCNT already takes into account direction.
+    motorPosSpeedObject.thetaRaw = EQep1Regs.QPOSCNT;
+
+    // Check an index occurrence
+    if (EQep1Regs.QFLG.bit.IEL == 1) EQep1Regs.QCLR.bit.IEL = 1;   // Clear interrupt flag
+
+    //Transform into electrical angle [rad]
+    motorPosSpeedObject.thetaElec = motorPosSpeedObject.thetaRaw * POLE_PAIRS * REV_TO_RAD * ENCODER_STEPS_INVERSE;
+    if (motorPosSpeedObject.thetaElec >= TWO_PI) motorPosSpeedObject.thetaElec -= TWO_PI;
+    motorPosSpeedObject.thetaMech = motorPosSpeedObject.thetaRaw * REV_TO_RAD * ENCODER_STEPS_INVERSE;
+
+    motorPosSpeedObject.speedTempCount += SW_PERIOD_US;
+}
+
+
+/*
+ *  To ensure the maximum accuracy in the position, the period between two measurements is minimized to switching frequency.
+ *  However, the rotor speed cannot be calculated at the same frequency since it will lead to errors.
+ *  Instead, the speed will be calculated every 10 degrees.
+ */
+void motorSpeedCalc(void)
+{
+    float deltaTheta;
+
+    deltaTheta = obtainDeltaTheta();
+
+    //Speed is only calculated every 10 degrees.
+    if (deltaTheta > DEG_10_TO_RAD)
+    {
+        updateSpeed(deltaTheta);
+        calcOtherSpeeds(); //Calculate other speeds.
+    }
+}
+
+
+float readRotorElecAngleRad(void)
+{
+    return motorPosSpeedObject.thetaElec;
+}
+
+float readRotorElecFreqHz(void)
+{
+    return motorPosSpeedObject.freqElec;
+}
+
+float readRotorMechAngleDeg(void)
+{
+    return motorPosSpeedObject.thetaMech;
+}
+
+float readRotorMechFreqHz(void)
```

Comment:
this function has a different name than its declaration, have you compiled?

Reply:
True, it was compiling even though it wasn't the same, thanks!